libs/contrib/Decidable/Finite/Fin.idr

```diff
@@ -0,0 +1,91 @@
+module Decidable.Finite.Fin
+
```

Comment:
This seems way too complicated. I expect that a simple:

* either `P FZ` holds
* or we use the fact that `P . FS` is decidable

ought to work (the size of `n` is structurally decreasing).

Cf. a crude attempt at deciding the universal quantifier in Agda
here: https://github.com/agda/agda-stdlib/issues/1181

Reply:
I might be missing something obvious, but that seems fundamentally different than what I'm doing, because in your version, `P` is a predicate on Nats, where in this one, `P` is a predicate on `Fin n` for *a fixed n*, which is where all the complication comes from (since you get a value of a smaller type when pattern matching on  Fin).

Now that I think about it, I guess we can prove the more general version for `P : Nat -> Type`, and for any `P' : (Fin n) -> Type`, we could call the Nat version with `\x => (lte : LT x n ** P' (natToFinLTE x lte) )`. Which kind of makes me question the value of Fin overall vs `Fin' n = (x : Nat ** LT x n)`, since it would be 

Alright, I'll try revising it to the `Nat` version. Thanks for the feedback!

Reply:
Here is what I meant

```idris
dec : (n : Nat) -> {0 p : Fin n -> Type} ->
      (pdec : (k : Fin n) -> Dec (p k)) ->
      Dec (k ** p k)
dec Z pdec = No (absurd . fst)
dec (S n) pdec =
  case pdec FZ of
    Yes pz => Yes (FZ ** pz)
    No npz => case dec n (\ k => pdec (FS k)) of
      Yes (k ** pk) => Yes (FS k ** pk)
      No npk => No $ \case
        (FZ ** pz) => absurd (npz pz)
        (FS k ** pk) => absurd (npk (k ** pk))
```

Reply:
... oh. Yeah, I hadn't thought of doing `(\ k => pdec (FS k))` in the recursive call. That is much better. I'll put it all together and revise the PR.